### ISSUE TITLE
service reflection: downloading transitive closure for a given symbol may not work

### ISSUE THREAD

**Issue by jhump** [grpc/grpc#19832]
In a dynamic client that uses service reflection to introspect on the server's schema, the existing operations are insufficient.

In order to reconstruct the schema in the client, like for creating descriptors in the client that can be used to create a dynamic stub and dynamic messages, we have to download the entire transitive closure for a particular service (all of the files it imports and the files they import and so on).

However, the only way provided to do that is to ask for files by name. However, trying to link files only by name is brittle and error-prone. If an import statement uses a relative path when referring to a dependency that is _different_ than the relative path used to compile that dependency (in the event they were compiled with two different invocations of `protoc`), then linking fails.

This root issue is the same as in https://github.com/golang/protobuf/issues/567 (which is the same problem, but just trying to link descriptors that are compiled into a Go program via the Go code generated by `protoc`).

The `FileDescriptorResponse`, which can include multiple descriptors, should probably also include information to assist with mapping import statements in one file (e.g. `FileDescriptorProto.dependency`) with the target file's _actual_ name (e.g. `FileDescriptorProto.name`). A discrepancy between these two is how linking breaks.

The Go implementation does not even bother sending an entire transitive closure in `FileDescriptorResponse` messages; it only sends back descriptors one at a time, requiring callers to then follow-up to request files by name based on the imports in the one file (and then doing so recursively to get the entire set of transitive dependencies). It would be great if the spec could be tightened up, possibly requiring that an entire set be supplied instead of just a single file and also that it _must_ supply the mappings if the set of files contains any such discrepancies.

A possible way to model this mapping in proto:

```proto
// Serialized FileDescriptorProto messages sent by the server answering
// a file_by_filename, file_containing_symbol, or file_containing_extension
// request.
message FileDescriptorResponse {
  // Serialized FileDescriptorProto messages. We avoid taking a dependency on
  // descriptor.proto, which uses proto2 only features, by making them opaque
  // bytes instead.
  repeated bytes file_descriptor_proto = 1;

  // Mappings to support linking files when the dependency name in one file
  // does not match the target file's name. This is index-correlated with the
  // above field. So the first entry in this field corresponds to imports for
  // the first descriptor in the above field. This field may have fewer
  // entries than the above field, which means that descriptors that have no
  // corresponding mapping need no such mappings to correctly link.
  repeated ImportMap = 2;
  message ImportMap {
    // This maps the dependency for a particular file, keyed by index in that
    // file descriptor's dependency field, to an actual descriptor. The value
    // is the index for the descriptor in the enclosing response's
    // file_descriptor_proto field.
    map<int32, int32> import_map = 1;
  }
}
```

**Comment by jhump** [comment]
Another solution, that does not require adding any import mapping fields to the current messages, is to strengthen the spec for `FileDescriptorResponse`. Instead of saying it *may* include the transitive closure, it could _require_ the transitive closure and also state the precise order so that a client could correctly link the files by just finding a given dependency at a computed index in the list of descriptors.

Of course, this is at odds with the text that states a descriptor may be omitted if it has already been sent to the client. So the response message could allow an _empty_ descriptors (zero bytes), which would indicate that the client should have already received the file. However, it should never do so for imports where the actual import statement used a different path/name than present in the target file descriptor itself (otherwise, we're back to the same problem trying to link by name).

**Comment by gnossen** [comment]
I don't believe that reflection currently has an owner, but in order to get some eyes on this...
CC @markdroth @ejona86 @dfawley

**Comment by ejona86** [comment]
@dklempner

I don't know what you mean by relative imports. Relative imports aren't a thing in protoc from what I can tell. Are you adding passing protoc a `-I` argument for each subfolder?

> If an import statement uses a relative path when referring to a dependency that is different than the relative path used to compile that dependency

I don't really understand that, because the descriptors are basically exactly what was passed from protoc to the protoc plugin. There's only one step involved: "parse the .proto into a descriptor."

**Comment by jhump** [comment]
> Relative imports aren't a thing in protoc

@ejona86, this did not mean "relative to the current proto file". I think, in fact, that _all_ import statements are relative: `protoc` will complain if it sees an absolute path. Anyhow, the description above goes into a little more detail. But the gist is: if you invoke `protoc` more than once -- once to compile file A and a different time to compile B, which imports A -- and you use _different_ relative path for A in invocation 1 compared to the import path used in invocation 2 (which, yes, requires different proto include paths), you encounter this problem. This is actually very common when the source and the importer are in different repos. Since proto does not have any official way to share sources/build across repos, most people have their own solution of copying files and/or using git submodules/subtrees. This often means that the proto path and thus import path for the importer are different than the original invocation of protoc used to compile the dependency.

Some more details and a simple concrete example can be found here: https://github.com/golang/protobuf/issues/567

While that issue is specifically about linking descriptors that are compiled into a Go package, the same problem exists for a client that is trying to download those same descriptors and link them locally. (When I say "link", I mean build a "rich" descriptor from the transitive closure of descriptor protos, creating a fully cross-linked structure, which makes the descriptors usable.)

**Comment by dklempner** [comment]
Regarding strengthening the spec: the spec (at least server-reflection.md) **does** require returning transitive imports. This protocol is almost unusably chatty (at least for internal-Google programs which often have sprawling proto dependencies) if it does not.

The server reflection proto is somewhat silent on the topic but retains language about how servers are allowed to omit protos which have already been returned, which is essentially meaningless unless these queries are transitive.

**Comment by jhump** [comment]
> Regarding strengthening the spec: the spec (at least server-reflection.md) does require returning transitive imports.

@dklempner, does that mean this other issue I filed is really a bug, not an enhancement request?
https://github.com/grpc/grpc-go/issues/2949

So if the spec were strengthened to strictly describe the order of files, then imports could be correctly linked even if there is a conflict between the names used to import the file and to compile it, which would totally solve this problem.

However, such a strengthening alone won't work because it also says that some descriptors may be omitted if they have already been sent to the client. But there's no way, in the current response structure, to indicate _which_ descriptors are being omitted. If they _have_ been sent to the client, the only way the client could cache them realistically is by name, but then we're back to the same problem, in the event that an import statement does not match that name.

Thinking about this more, the whole idea of being able to omit descriptors if they've already been sent is brittle and subject to the same problems: the only way a client currently has to identify a file is by name. But what if two `foo.proto` files, in different directories, are compiled using the relative path `foo.proto`? How would the client know which is intended by a particular downloaded file when it sees an import name of "foo.proto"? It could not reliably fulfill all operations with cached file descriptors because of this confusion.

**Comment by ejona86** [comment]
> But the gist is: if you invoke protoc more than once -- once to compile file A and a different time to compile B, which imports A -- and you use different relative path for A in invocation 1 compared to the import path used in invocation 2 (which, yes, requires different proto include paths), you encounter this problem

You're saying the same proto file has two different names? Yeah, that seems unsupported. You can't go around willy-nilly changing the proto file name, in the same way you can't go around changing the proto package name. Bringing in @acozzette from protobuf team for more authoritative response.

> This is actually very common when the source and the importer are in different repos. Since proto does not have any official way to share sources/build across repos, most people have their own solution of copying files and/or using git submodules/subtrees. 

It seems like there should be some judicial use of `-I` to make the import paths the same.

**Comment by acozzette** [comment]
Yes, for better or worse there's a fundamental assumption that each .proto file has just one path. If different protoc invocations consider a file to be at different logical paths then that will cause problems.

**Comment by jhump** [comment]
@acozzette, @ejona86, that's easy enough in a monorepo or even within a single organization. But if gRPC is going to increase as a thing for _external_ APIs, like Google is doing with its cloud services, there needs to support for cross-repo and cross-organization usages, which is where it's trivial for this kind of problem to pop up.

I think it's a real issue is that it's not _specified_ anywhere that this is required. And it if _is_ required, there's nothing in `protoc` that attempts to validate it (no way in the language for a file to declare what its name should be).

So if I need to use some 3rd party proto, nowhere is it documented what the import path (and thus `-I` parameter) _should_ be. It's buried inside of generated code, in the name field of an embedded&encoded descriptor proto.

Some repos I've seen actually have good organization of files, but then a Makefile that just runs `protoc` inside of each directory, using just the files' base names (and including an `-I .` to make it work). Needless to say, that makes for very difficult files to consume in a way that will actually work with reflection.

There are no warnings from `protoc` and not even anything in the documentation to suggest that this is bad idea, so it happens. But it totally breaks reflection. And the generated code works (unless there were a test for reflection), so it goes unnoticed.

**Comment by ejona86** [comment]
@jhump, I think it is pretty easy to make it work with the restriction in general, but I agree that unless people know about the restriction they won't necessarily set things up to work correctly. This probably isn't too much of an issue with Java, due to our involvement with the Maven/Gradle plugins and consuming .protos from with .jars. But other languages would be trivially impacted.

I've created https://github.com/protocolbuffers/protobuf/issues/6492 to trace the documentation/verification issue. I've commented on the go issue stating it seems like a bug, because of the language in server-reflection.md. I don't think there's really a way to fix the reflection service, because in this environment the descriptor information is inconsistent.
